mfa_duo.go

```diff
@@ -0,0 +1,86 @@
+package main
+
+import (
+	"net/http"
+	"time"
+	"strings"
+	"github.com/duosecurity/duo_api_golang"
+	"github.com/duosecurity/duo_api_golang/authapi"
+	yaml "gopkg.in/yaml.v2"
+)
+
+func init() {
+	registerMFAProvider(&mfaDuo{})
+}
+
+type mfaDuo struct{
+	Ikey  string `yaml:"ikey"`
+	Skey string `yaml:"skey"`
+	Host string `yaml:"host"`
+	UserAgent string `yaml:"userAgent"`
+}
+
+// ProviderID needs to return an unique string to identify
+// this special MFA provider
+func (m mfaDuo) ProviderID() (id string) { return "duo" }
+
+// Configure loads the configuration for the Authenticator from the
+// global config.yaml file which is passed as a byte-slice.
+// If no configuration for the Authenticator is supplied the function
+// needs to return the errProviderUnconfigured
+func (m *mfaDuo) Configure(yamlSource []byte) (err error) {
+	envelope := struct {
+		MFA struct {
+			Duo *mfaDuo `yaml:"duo"`
+		} `yaml:"mfa"`
+	}{}
+
+	if err := yaml.Unmarshal(yamlSource, &envelope); err != nil {
+		return err
+	}
+
+	if envelope.MFA.Duo == nil {
+		return errProviderUnconfigured
+	}
+
+	m.Ikey = envelope.MFA.Duo.Ikey
+	m.Skey = envelope.MFA.Duo.Skey
+	m.Host = envelope.MFA.Duo.Host
+	m.UserAgent = envelope.MFA.Duo.UserAgent
+	return nil
+}
+
+// ValidateMFA takes the user from the login cookie and performs a
+// validation against the provided MFA configuration for this user
+func (m mfaDuo) ValidateMFA(res http.ResponseWriter, r *http.Request, user string, mfaCfgs []mfaConfig) error {
+	var keyInput string
+	// Look for mfaConfigs with own provider name
+	for _, c := range mfaCfgs {
+		if c.Provider != m.ProviderID() {
+			continue
+		}
+		ip := r.Header.Get("X-Real-IP")
```

Comment:
This should not rely only on `X-Real-IP` but check also `X-Forwarded-For` etc. Also the login is requested directly through the browser: The reported remote-IP might be fine.

See https://github.com/Luzifer/go_helpers/blob/master/http/logHandler.go#L47-L56

Reply:
Rename: `remoteIP`

Reply:
My aim here was to extract the IP of the remote user as duo can also use this to auto allow auth if required, i.e if the request came from an internal subnet it could just pass auth without requiring the MFA factor.

Im unsure what to change this too as, im assuming nginx would pass the remote address from 
`proxy_set_header X-Real-IP $remote_addr;`

Reply:
That's understood though I disagree with disabling MFA for specific IP sources (perimeter security tends to be very insecure and break quite fast).

About the IP: Do not assume the administrator set this header. There needs to be a fallback for it: The remote-addr. Also there are multiple common headers for transmitting forwarded IPs (`X-Forwarded-For`, `X-Real-IP`).

Therefore try to resolve them and if this fails provide the remote-addr Go can see. (Like in the linked code)

If you want I can add this to the PR.

Reply:
I agree that disabling MFA for specific IP sources is a bad idea, I was just aiming to allow all of duos features to work. This is  also something you would configure on the duo side so having this in would still need the duo ACL to be configured to look at the IP.

Please do add this to the PR for me as im unsure how you want this implemented, the important thing to note is duo can only be passed one IP per auth and this ip should ideally be end users ip.